packages/cli/src/modules/data-store/data-store-column.repository.ts

```diff
@@ -0,0 +1,141 @@
+import { DataStoreCreateColumnSchema } from '@n8n/api-types';
+import { CreateTable, DslColumn } from '@n8n/db';
+import { Service } from '@n8n/di';
+import { DataSource, EntityManager, Repository } from '@n8n/typeorm';
+import { UserError } from 'n8n-workflow';
+
+import { DataStoreColumnEntity } from './data-store-column.entity';
+import { addColumnQuery, deleteColumnQuery, toDslColumns, toTableName } from './utils/sql-utils';
+
+@Service()
+export class DataStoreColumnRepository extends Repository<DataStoreColumnEntity> {
+	constructor(dataSource: DataSource) {
+		super(DataStoreColumnEntity, dataSource.manager);
+	}
+
+	async getColumns(rawDataStoreId: string, em?: EntityManager) {
+		const executor = em ?? this.manager;
+		const columns = await executor
+			.createQueryBuilder(DataStoreColumnEntity, 'dsc')
+			.where('dsc.dataStoreId = :dataStoreId', { dataStoreId: rawDataStoreId })
+			.getMany();
+
+		// Ensure columns are always returned in the correct order by columnIndex,
+		// since the database does not guarantee ordering and TypeORM does not preserve
+		// join order in @OneToMany relations.
+		columns.sort((a, b) => a.columnIndex - b.columnIndex);
+
+		return columns;
+	}
+
+	async addColumn(dataStoreId: string, schema: DataStoreCreateColumnSchema) {
+		return await this.manager.transaction(async (em) => {
+			const existingColumnMatch = await em.existsBy(DataStoreColumnEntity, {
+				name: schema.name,
+				dataStoreId,
+			});
+
+			if (existingColumnMatch) {
+				throw new UserError(
+					`column name '${schema.name}' already taken in data store '${dataStoreId}'`,
+				);
+			}
+
+			if (schema.columnIndex === undefined) {
+				const columns = await this.getColumns(dataStoreId, em);
+				schema.columnIndex = columns.length;
+			} else {
+				await this.shiftColumns(dataStoreId, schema.columnIndex, 1, em);
+			}
+
+			const column = em.create(DataStoreColumnEntity, {
+				...schema,
+				dataStoreId,
+			});
+
+			await em.insert(DataStoreColumnEntity, column);
+
+			// Create user table if it does not exist
+			const tableName = toTableName(dataStoreId);
+			const queryRunner = em.queryRunner;
+			if (!queryRunner) {
+				throw new Error('QueryRunner is not available');
+			}
+			const tableExists = await queryRunner.hasTable(tableName);
+			if (!tableExists) {
+				const dslColumns = [
```

Comment:
Isn't this a responsibility of the data store repository? I assume trying to add a column for non existent table is not expected so if that's true adding the table back would also be not expected.